Travis setup

jb803 opened the pull request:
Added everything required for Travis CI to automatically build and test the code. I've assumed we want to stick with installing FEniCS via anaconda. 

Each Travis job:

1. Downloads miniconda
2. Installs FEniCS, pytest, sphinx and black under the specified python version (just 3.6 at the moment)
3. Runs the relevant test suite.

Tox doesn't play well with anaconda so there's now a test script, `provision/travis/run-test.sh` , which runs different tests based on the `TEST_ENV` environment variable. If `TEST_ENV` is not passed then it will run all the tests. 

The current tests are:
1. `pytest` on the `test/` folder. I've filtered out `DepreciationWarning` and `PendingDepreceationWarning` because it turns out `import dolfin` causes about a billion of these to be produced (who would have thought it? :p)
2. Tests building the doucmentation via sphinx
3. `black --check tonic/`  . Checks to make sure `black` is happy with all the module code in `tonic/`. I don't know if we want to add `test/` to the files checked by `black`.  What do people think?

You don't seem to be able to get activate Travis on a repository until the `.travis.yml` file is present in the master branch. When this is merged, I'll enable Travis and then make the Travis checks a requirement for merging pull requests.

Commit:
Initial pass at setting up Travis CI and coveralls

Tox is known to behae badly with anaconda. Where tox would normally be used
I've substituted a bash script. To simulate what the TavisCI agent will run use
./provision/travis/run-tests.sh

It would be good to find a tox alternative. (Or failing that we could use a different method
of installing dependencies e.g. docker or using the apt-get version of FENICS...)

Commit:
Fixed typo in .travis.yml

Commit:
Updated anaconda environment name

Commit:
We now install all dependencies at the same time.

We now install all dependencis in the same command. This reduces
time waiting for anaconda to solve the environment

Commit:
Added sphinx dependency

Commit:
Updated README with details of test script

Saintis commented on .gitignore:
```diff
@@ -4,3 +4,7 @@
+#ignore coverage info
+.coverage
+#ignore pytest chace
+.pytest_cache/*
```
Needs line break

Saintis commented on .travis.yml:
```diff
@@ -0,0 +1,48 @@
+script: provision/travis/run-tests.sh
+after_success:
+  - coveralls
+
```
way too many line breaks

Saintis commented on provision/travis/run-tests.sh:
```diff
@@ -0,0 +1,23 @@
+     black --check tonic/
+else
+    echo Uknown test environment
+fi
```
needs line break

Saintis commented on provision/travis/run-tests.sh:
```diff
@@ -0,0 +1,23 @@
+elif [ "$TEST_ENV" = "doc" ]; then
+    sphinx-build -b html doc/source/ doc/build/
+elif [ "$TEST_ENV" = "black" ]; then
+     black --check tonic/
```
some repeated lines of code, is it possible to rewrite this so the sphinx, pytest and black lines are only written once?

Saintis replied:
also do black on whole project, including tests.

jb803 replied:
Test invocations are declared at beginning of the bash file now.
black runs also checks the test/ directory

Saintis commented on test/test_deps.py:
```diff
@@ -0,0 +1,4 @@
+"""Tests  various dependencies. This is an initial example test required to allow pytest to be happy when there are no actual tests written """
+
+def test_importDolfin():
+    import dolfin as dolf
```
needs line break

Commit:
Fixed line breaks

Commit:
Reformmated test script

1. Removed repetition of test invocations
2. Test will now fail if an unknwon test environment is specified (previously it would silently succeed)

Commit:
Fixed typo in test error message
